### DISCUSSION TITLE
Esp32 Audio Kit as SnapClient

### DISCUSSION THREAD

**Discussion by unknown0816** [pschatzmann/arduino-snapclient#8] [Q&A] [answered]
I discovered this project today and I was quiet happy, because this is exactly what I want to do. So thank you for sharing this!

So let me start by describing what I want to do:
* I have a working snap-server and multiple snap-clients.
* I brought a esp32 audio-kit (I guess from AI) and want to connect this with the snap-server as another snap-client.

To get used to the esp32 audio kit I tested the Bluetooth example and it worked fine. I connected my phone with the audio-kit and I heart the sound via the speakers connected to the audio kit.

So after that I found this example, which sound pretty much as my goal: https://github.com/pschatzmann/arduino-snapclient/blob/main/examples/SnapClientAudioKit/SnapClientAudioKit.ino 

I changed the codec to Flac. So changed the include, initialized the FLACDecoder flac and added it to the SnapClient init. Further, I changed the SnapClient Config to find WIFI and the SnapServer.

After flashing I can see the Arduino SnapClient in the SnapServer Monitoring. But sadly, I do not hear anything.

Does somebody know what the problem could be?

Update:
I looked into the Serial Monitor and I found two strange things:
* Warning Pin '13' not set up because of conflict
* [E] CodecFlac.h : 108 - not active

The last error is printed on each sound-packet as far as I can see.

**Comment by pschatzmann** [comment, marked as answer]
Maybe I messed up something with the FLAC decoder in one of my recent changes. I suggest that you try with my recommended Opus

**Reply by unknown0816** [reply]
Okay. I can try to get Opus work. In a first try even my existing setup did not work anymore. But I will look deeper into it.

btw: I did change the LogLevel and can provide a more detailed Log: 
[log.txt](https://github.com/pschatzmann/arduino-snapclient/files/14447856/log.txt)


What I noticed:
* After the init, there is some kind of loop from CodecFLAC.h line 149 to line 157. This is depending on the define: FLAC_READ_TIMEOUT_MS.
* After that the "is_active" flag is set to false and the next loop starts which corresponds with the audio packets from AudioEncoded.h line 358 to line 374

**Reply by unknown0816** [reply]
Okay. I don't know what I did before. Opus works with the existing setup, too.
And it works with the Audio-Kit, too. But the sound quality is pretty bad. Sometime there are small breaks or it sounds like the audio kit can't handle the amount of data and just skips parts or similar.
And I can imagine this would be even worse with flac.

**Reply by pschatzmann** [reply]
Then try to use the SnapProcessorRTOS

**Reply by pschatzmann** [reply]
or increase the i2s buffer

Maybe you should also read https://github.com/pschatzmann/arduino-snapclient/issues/7

**Reply by unknown0816** [reply]
Thank you for the answers.

However, I tried another snapclient project that uses the esp adf instead of arduino. After I changed some PINs and minor volume controls this worked quiet well and I don't have to change the snapserver streams from FLAC to OPUS.

If anybody is interested:
* [The Snapclient](https://github.com/CarlosDerSeher/snapclient)
* [PIN change for AI Thinker Audio Kit](https://github.com/Ai-Thinker-Open/ESP32-A1S-AudioKit)

**Comment by pschatzmann** [comment]
I agree: I you just look for a easy working solution for one of the  supported ESP32 based Audio Boards with PSRAM this is definitly the better solution. 

However, the goal of this project was to provide a platform independent implementation with minimal RAM requirements. This requires considerable more time for the user to fine tune things. But it comes with the advantage that you can integrate this into your own application.

I will update the readme to make this clear.
